### PR TITLE
[TT-5279] Shared type, extension orphan and validation clean up

### PR DESCRIPTION
**Pull request by Aenimus**
[changelog]
internal: Shared types cannot be extended; added unique union member validation; unresolved extension orphans in the supergraph return an error

### CODE REVIEW

File: pkg/federation/sdlmerge/check_extension_orphans.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	walker.RegisterEnterEnumTypeExtensionVisitor(p)
+	walker.RegisterEnterInputObjectTypeExtensionVisitor(p)
+	walker.RegisterEnterInterfaceTypeExtensionVisitor(p)
+	walker.RegisterEnterObjectTypeExtensionVisitor(p)
+	walker.RegisterEnterScalarTypeExtensionVisitor(p)
+	walker.RegisterEnterUnionTypeExtensionVisitor(p)
+}
+
+func (p *checkUnresolvedExtensionOrphansVisitor) EnterDocument(operation, _ *ast.Document) {
+	//operation.DeleteRootNodes(operation.Index.MergedTypeExtensions)
```
Comment (Aenimus) [inline pkg/federation/sdlmerge/check_extension_orphans.go:35]: Will delete this

File: pkg/federation/sdlmerge/sdlmerge_test.go [hunk 2]
```diff
@@ -99,15 +103,15 @@ func TestMergeSDLs(t *testing.T) {
 		accountSchema, productSchema, reviewSchema, likeSchema, disLikeSchema, paymentSchema, onlinePaymentSchema, classicPaymentSchema,
 	))
 
-	t.Run("should merge product and review sdl and leave `extend type User` in the schema", runMergeTest(
-		productAndReviewFederatedSchema,
-		productSchema, reviewSchema,
-	))
+	//t.Run("should merge product and review sdl and leave `extend type User` in the schema", runMergeTest(
```
Comment (Aenimus) [inline pkg/federation/sdlmerge/sdlmerge_test.go:106]: Not sure how to handle this. Why did Cherom think it should be valid?

File: pkg/federation/sdlmerge/object_type_extending.go [hunk 3]
```diff
@@ -23,15 +26,38 @@ func (e *extendObjectTypeDefinitionVisitor) EnterDocument(operation, _ *ast.Docu
...
 func (e *extendObjectTypeDefinitionVisitor) EnterObjectTypeExtension(ref int) {
-	nodes, exists := e.operation.Index.NodesByNameBytes(e.operation.ObjectTypeExtensionNameBytes(ref))
+	nameBytes := e.operation.ObjectTypeExtensionNameBytes(ref)
+	nodes, exists := e.operation.Index.NodesByNameBytes(nameBytes)
 	if !exists {
 		return
 	}
 
+	hasExtended := false
+	shouldReturn := isQueryOrMutation(nameBytes)
```
Comment (Aenimus) [inline pkg/federation/sdlmerge/object_type_extending.go:36]: I'm not sure how else we avoid tripping over queries and mutations (and whether there are other edge cases), but here is a possible solution.

File: pkg/astvalidation/rule_populated_type_bodies.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+func (p *populatedTypeBodiesVisitor) EnterObjectTypeExtension(ref int) {
+	definition := p.definition
+	if !definition.ObjectTypeExtensions[ref].HasFieldDefinitions {
+		p.Report.AddExternalError(operationreport.ErrTypeBodyMustNotBeEmpty("object extension", definition.ObjectTypeExtensionNameString(ref)))
+		return
+	}
+}
+
+func isRootType(nameBytes []byte) bool {
```
Comment (Aenimus) [inline pkg/astvalidation/rule_populated_type_bodies.go:119]: This is a temporary solution

File: pkg/astvalidation/definition_validation.go [hunk 5]
```diff
@@ -51,3 +53,20 @@ func (d *DefinitionValidator) Validate(definition *ast.Document, report *operati
...
 	return Valid
 }
+
+func IsRootType(nameBytes []byte) bool {
+	length := len(nameBytes)
+	return isQuery(length, nameBytes) || isMutation(length, nameBytes) || isSubscription(length, nameBytes)
+}
+
+func isQuery(length int, b []byte) bool {
+	return length == 5 && b[0] == 'Q' && b[1] == 'u' && b[2] == 'e' && b[3] == 'r' && b[4] == 'y'
```
Comment (pvormste) [inline pkg/astvalidation/definition_validation.go:63]: I would suggest to create some constants in the `ast` package like this:
```go
const DefaultQueryTypeName = "Query"
const DefaultMutationTypeName = "Mutation"
const DefaultSubscriptionTypeName = "Subscription"
```

... and use those for comparison. You could also use them to remove some hardcoded values in `ast` package:
https://github.com/TykTechnologies/graphql-go-tools/blob/master/pkg/ast/ast_root_operation_type_definition.go#L54

File: pkg/astvalidation/rule_populated_type_bodies.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	object := definition.ObjectTypeDefinitions[ref]
+	switch object.HasFieldDefinitions {
+	case true:
+		if IsRootType(nameBytes) {
+			return
+		}
+		for _, fieldRef := range definition.ObjectTypeDefinitions[ref].FieldsDefinition.Refs {
+			fieldNameBytes := definition.FieldDefinitionNameBytes(fieldRef)
+			length := len(fieldNameBytes)
+			if length < 2 || fieldNameBytes[0] != '_' || fieldNameBytes[1] != '_' {
```
Comment (pvormste) [inline pkg/astvalidation/rule_populated_type_bodies.go:107]: maybe it makes sense to create a helper function for this, there is also another rule checking for reserved type names: https://github.com/TykTechnologies/graphql-go-tools/blob/master/pkg/astvalidation/rule_unique_field_definition_names.go#L119

(maybe there are even more?)

**Review by pvormste — CHANGES_REQUESTED** [review]

**Review by pvormste — APPROVED** [review]

File: pkg/federation/sdlmerge/interface_type_extending.go [hunk 7]
```diff
@@ -3,35 +3,48 @@ package sdlmerge
...
 }
 
 func (e *extendInterfaceTypeDefinitionVisitor) EnterDocument(operation, _ *ast.Document) {
-	e.operation = operation
+	e.document = operation
 }
 
 func (e *extendInterfaceTypeDefinitionVisitor) EnterInterfaceTypeExtension(ref int) {
-	nodes, exists := e.operation.Index.NodesByNameBytes(e.operation.InterfaceTypeExtensionNameBytes(ref))
+	document := e.document
```
Comment (spetrunin) [inline pkg/federation/sdlmerge/interface_type_extending.go:29]: a bit unclear why do u need this var?

Comment (Aenimus) [reply pkg/federation/sdlmerge/interface_type_extending.go:29]: I call e.document 3 or 4 times

**Review by spetrunin — APPROVED** [review]

**Pull request merged**